README.md
## Http ##

High-level HTTP client and (Slim) service utility.  
Client provides response validation and mocking.  
Service provides standardized and predictable responses.

<details>
<summary>

### ▹ Client ###
</summary>

#### Request options ####

The ```HttpClient->request()``` method accepts it's own options  
as well as options for the underlying [RestMini Client](https://github.com/simplecomplex/restmini#client-options).

##### Own options #####

- (bool) **debug_dump**: log request and response
- (bool|arr) **cacheable**: cache response/load response from [cache](https://github.com/simplecomplex/php-cache)
- (bool|arr) **validate_response**: validate the response body against a [validation rule set](https://github.com/simplecomplex/php-validate)
- (bool|arr) **mock_response**: don't send the request, return predefined mock response
- (int) **retry_on_unavailable**: millisecs; try again later upon  
     503 Service Unavailable|host not found|connection failed
- (arr) **require_response_headers**: list of response header keys required
- (bool) **err_on_endpoint_not_found**: err on 404 + HTML response
- (bool) **err_on_resource_not_found**: err on 204 or 404 + JSON response
- (arr) **log_warning_on_status**: key is status code, value is true

##### Options processing #####

Apart from options passed directly to ```HttpClient->request()```  
there may also exist settings (see [Config](https://github.com/simplecomplex/php-config)) for the:
- **provider**: the service host
- **service**: a group of endpoints
- **endpoint**: the actual endpoint
- **method**: literal HTTP method or an alias (like GET aliases index and retrieve)

During request preparations settings and options get merged, so that  
– _options_ override _method_ settings, which override _endpoint_ settings, which override... (you get it).  
Sounds like a lot of work, but it isn't really.


  <details>
  <summary>
  
  #### ▹▹ Client CLI commands ####
  </summary>

##### (remote) Service configuration #####

```bash
# Show provider settings.
php cli.php config-get -a global http-provider.prvdr

# Show service settings.
php cli.php config-get -a global http-service.prvdr.example-service

# Show endpoint settings.
php cli.php config-get -a global http-endpoint.prvdr.example-service.ndpnt

# Show method settings.
php cli.php config-get -a global http-method.prvdr.example-service.ndpnt.GET
```

##### Validation rule sets #####

```bash
# Show cached validation rule set.
php cli.php cache-get http-response_validation-rule-set prvdr.example-service.ndpnt.GET

# Delete cached validation rule set.
php cli.php cache-delete http-response_validation-rule-set prvdr.example-service.ndpnt.GET
```

##### Mock responses #####

```bash
# Show cached mock response.
php cli.php cache-get http-response_mock prvdr.example-service.ndpnt.GET

# Delete cached mock response.
php cli.php cache-delete http-response_mock prvdr.example-service.ndpnt.GET
```
  </details>

  <details>
  <summary>
  
  #### ▹▹ Client error codes ####
  </summary>
  
  For every error code there's an equivalent prefab safe and user-friendly (localizable) error message.
  
  - ```unknown```: overall error fallback
  - ```local-unknown```: local error fallback
  - ```local-algo```: in-package logical error
  - ```local-use```: invalid argument et al.
  - ```local-configuration```: bad config var
  - ```local-option```: bad option var
  - ```local-init```: RestMini Client or cURL cannot request
  - ```host-unavailable```: DNS or actually no such host
  - ```service-unavailable```: status 503 Service Unavailable
  - ```too-many-redirects```: too many redirects
  - ```timeout```: cURL 504
  - ```timeout-propagated```: status 504 Gateway Timeout
  - ```response-none```: cURL 500 (RestMini Client 'response-false')
  - ```remote```: status 500 Internal Server Error
  - ```remote-propagated```: remote says 502 Bad Gateway
  - ```malign-status-unexpected```: unsupported 5xx status
  - ```endpoint-not-found```: status 404 + Content-Type not JSON (probably HTML); no such endpoint
  - ```resource-not-found```: status 204, status 404 + Content-Type JSON; no such resource (object)
  - ```remote-validation-bad```: 400 Bad Request 
  - ```remote-validation-failed```: 412 Precondition Failed, 422 Unprocessable Entity
  - ```response-type```: content type mismatch
  - ```response-format```: parse error
  - ```benign-status-unexpected```: unsupported non-5xx status
  - ```header-missing```: setting/option _require_response_headers_
  - ```response-validation```: response body validation failure; service will send X-Http-Response-Invalid header
  
  </details>

</details>

<details>
<summary>

### ▹ Service ###
</summary>

Producing a service response is not as hard as requesting a remote service,  
so the service part of **Http** is not as rich as the client part.  
It is assumed that one will simply echo something and send some headers
– or use a service framework like [Slim](https://www.slimframework.com/).

The ```HttpServiceSlim``` class suggest means of interacting with Slim, and **Http** includes a simple example.
@todo  
And **Http** also provides a few other service utilities.

#### Allowing Cross Origin requests ####

Preferably only at development site. Necessary when developing Angular-based frontend.

Place a ```.cross_origin_allow_sites``` text file in document root, containing list allowed sites, like:  
```http://localhost:4200,http://my-project.build.local.host:80```

  <details>
  <summary>
  
  #### ▹▹ Service error codes ####
  </summary>
  
  For every error code there's an equivalent prefab safe and user-friendly (localizable) error message.
  
  - ```unknown```: overall fallback
  - ```request-unacceptable```: (some kind of) bad request; ```HttpResponseRequestUnacceptable```
  - ```unauthenticated```: authentication (login) failure; ```HttpResponseRequestUnauthenticated```
  - ```unauthorized```: authorization (permission) failure; ```HttpResponseRequestUnauthorized```
  - ```request-validation```: request header/argument validation failure; ```HttpResponseRequestInvalid```
  - ```frontend-response-format```: frontend only; parse error
  - ```frontend-response-validation```: frontend only; response validation failure
  
  </details>

</details>

### Service and client combined ###

#### Standardized wrapped response body ####

A service requestor should never be in doubt whether a request to your service went alltogether well.  
```HttpClient->request()``` returns ```HttpResponse``` object:

- status ```int```: suggested status to send to requestor
- headers ```array```: suggested headers to send
- body ```HttpResponseBody```
  - **success** ```bool```
  - **status** ```int```: status received from remote service
  - **data** ```mixed|null```: that actual data (on success)
  - **message** ```string|null```: safe and user-friendly message (on failure)
  - **code** ```int```: error code (on failure), or optionally some other flag (on success)
- originalHeaders ```array```: headers received – _not_ to be sent to requestor
- validated ```bool|null```: whether the response was validated, and then the outcome

The general idea is to always send a response body containing _metadata_ about the procedings – success, status, code.  
And – on failure – a prefab safe and user-friendly message, which the client can use to inform the visitor  
(instead of just failing silently/ungracefully).

When exposing a service that does a remote request, the service should however have access to more detailed info
about the remote request/response  
– therefore the further wrapping in an object that suggests status and headers (et al.).

<details>
  <summary>
  
#### ▹ Service response headers ####
</summary>

**Http** uses a number of custom response headers, to flag stuff to the client.  
Some are issued by ```HttpClient``` upon every remote request (the original/final statuses).  
Others are only issued if a service uses/sends one of the prefab ```HttpResponse``` extensions.

- <sup>(int)</sup>```X-Http-Original-Status```: status received from remote service (or interpretated ditto)
- <sup>(int)</sup>```X-Http-Final-Status```: final status to be sent to client
- ```X-Http-Response-Invalid```: response validation failure; ```HttpResponseResponseInvalid```
- ```X-Http-Mock-Response```: ```HttpClient``` never called remote service; used prefab mock response
- ```X-Http-Request-Invalid```: request header/argument validation failure; ```HttpResponseRequestInvalid```
- ```X-Http-Request-Unacceptable```: (some kind of) bad request; ```HttpResponseRequestUnacceptable```
- ```X-Http-Request-Unauthenticated```: authentication (login) failure; ```HttpResponseRequestUnauthenticated```
- ```X-Http-Request-Unathorized```: authorization (permission) failure; ```HttpResponseRequestUnauthorized```

</details>

### Requirements ###

- PHP >=7.0
- [PSR-3 Log](https://github.com/php-fig/log)
- [SimpleComplex Utils](https://github.com/simplecomplex/php-utils)
- [SimpleComplex Cache](https://github.com/simplecomplex/php-cache)
- [SimpleComplex Config](https://github.com/simplecomplex/php-config)
- [SimpleComplex Locale](https://github.com/simplecomplex/php-locale)
- [SimpleComplex Validate](https://github.com/simplecomplex/php-validate)
- [SimpleComplex RestMini](https://github.com/simplecomplex/restmini)
- [SimpleComplex Inspect](https://github.com/simplecomplex/inspect)

#### Suggestions ####
- [Slim](https://github.com/slimphp/Slim) 


CHANGELOG.txt
SimpleComplex/Http 2.x, 2019-02-03
----------------------------------
- Standard prefab HTTP responses shan't promise in body.message
  that the incident got logged.
- Release 2.1.5.

SimpleComplex/Http 2.x, 2018-10-11
----------------------------------
- Do not force CORS Access-Control-Allow-Credentials header
  even when no authentication involved.
  And never send that darned header if OPITONS request.
- Release 2.1.4.

SimpleComplex/Http 2.x, 2018-09-25
----------------------------------
- Send Access-Control-Allow-Credentials header.
- Release 2.1.3.

SimpleComplex/Http 2.x, 2018-09-23
----------------------------------
- Service Slim use exception and error handlers of Utils.
- Service Slim support multiple HTTP methods for single endpoint.
- Release 2.1.2.

SimpleComplex/Http 2.x, 2018-09-15
----------------------------------
- Request support more (validation failure) statuses; 410, 411,
  413 - 417, 423, 424.
- Release 2.1.

SimpleComplex/Http 2.x, 2018-09-06
----------------------------------
- Support response validation on HTTP status 202; may produce body.

SimpleComplex/Http 2.x, 2018-08-30
----------------------------------
- Provide fallback non-essential dependency injection container items
  for HttpClient and classes it may effectively use.

SimpleComplex/Http 2.x, 2018-06-25
----------------------------------
- Utils PathList::clear() renamed; from reset().
- Release 2.0.1.

SimpleComplex/Http 2.x, 2018-06-24
----------------------------------
- Support and prefer finding validation and mocking .json files via config .ini
  lists http:responseValidationRuleSetPaths and http:responseMockPaths, instead
  of files symlinked/copied to conf dir.
- Release 2.0.

SimpleComplex/Http 2.x, 2018-06-23
----------------------------------
- New major version because new config regime may break things for users.
- Don't provide separate base and override config .inis; user must of own accord
  override (in prod/test environments) using custom .ini file.

SimpleComplex/Http 1.x, 2018-06-19
----------------------------------
- HttpRequest validate and mock use Utils PathList instead of it's predecessor.
- Release 1.2.

SimpleComplex/Http 1.x, 2018-06-17
----------------------------------
- HttpRequest validate and mock use Utils FileNonUniqueException.

SimpleComplex/Http 1.x, 2018-06-16
----------------------------------
- Example .ini and .json files must use ext ...example.

SimpleComplex/Http 1.x, 2017-11-22
----------------------------------
- Client support status 401 Unauthorized (interpreted as unauthenticated)
  and status 403 Forbidden (interpreted as unauthorized).
- Release 1.0.6.

SimpleComplex/Http 1.x, 2017-11-09
----------------------------------
- Option err_on_endpoint_not_found means 404+HTML; not 404!JSON.
- Release 1.0.5.

SimpleComplex/Http 1.x, 2017-10-24
----------------------------------
- Support for status 409 Conflict.
- Release 1.0.4.

SimpleComplex/Http 1.x, 2017-10-06
----------------------------------
- Provide PSR-3 Log context; to support providing JsonLog event correlationId.
- Release 1.0.3.

SimpleComplex/Http 1.x, 2017-10-05
----------------------------------
- Don't assume non-empty response content type header when status 404.
- Don't validate response body unless status 200/201.
- Release 1.0.2.

SimpleComplex/Http 1.x, 2017-09-25
----------------------------------
- Error code typo.
- Options explained in README.
- Release 1.0.1.

SimpleComplex/Http 1.x, 2017-09-24
----------------------------------
- Release 1.0.

SimpleComplex/Http 1.x, 2017-09-23
----------------------------------
- Facilitate extending major classes; moved overridable class constants
  to HttpSettings class, and HttpServiceSlim no longer extends HttpService.

SimpleComplex/Http 1.x, 2017-09-22
----------------------------------
- Config and locale-text no longer risk that dupe key scalar values become array
  Removed warnings from documentation.

SimpleComplex/Http 1.x, 2017-09-19
----------------------------------
- HttpServiceSlim set PHP warning/notice error handler.
  And fix that origin exception wasn't logged in an exception/error handler
  itself raised an exception.
- HttpForwardableResponseException bugs.

SimpleComplex/Http 1.x, 2017-09-18
----------------------------------
- Frontend service error message headers.
- HttpForwardableResponseException to handle service error that happen deep
  within the framework.

SimpleComplex/Http 1.x, 2017-09-17
----------------------------------
- Cache control response headers.

SimpleComplex/Http 1.x, 2017-09-16
----------------------------------
- Use standard method to send 403 Forbidden response, except when cause is real
  unauthorized.

SimpleComplex/Http 1.x, 2017-09-14
----------------------------------
- HttpService/HttpServiceSlim obsolete constructor param config removed.

SimpleComplex/Http 1.x, 2017-09-10
----------------------------------
- .cross_origin_allow_sites file renamed; from .access_control_allow_origin.
- Cross origin Access-Control-Allow-Origin header must only be returned when
  when such site requests, and only contain it's _own_ protocol, host and port;
  and port only if request header Origin contains port.

SimpleComplex/Http 1.x, 2017-09-03
----------------------------------
- Slim initialization generalized and moved to HttpServiceSlim.
- HttpServiceSlim is now the overall bootstrapper,
  of Slim and the simplecomplex framework.

SimpleComplex/Http 1.x, 2017-09-02
----------------------------------
- Mock and validation rule set caches allow long keys; length 128.

SimpleComplex/Http 1.x, 2017-08-31

src/HttpServiceSlim.php
<?php
/**
 * SimpleComplex PHP Http
 * @link      https://github.com/simplecomplex/php-http
 * @copyright Copyright (c) 2017-2018 Jacob Friis Mathiasen
 * @license   https://github.com/simplecomplex/php-http/blob/master/LICENSE (MIT License)
 */
declare(strict_types=1);

namespace SimpleComplex\Http;

use Slim\Container;
use Slim\App;
use Slim\Http\Request;
use Slim\Http\Response;

use SimpleComplex\Utils\Dependency;
use SimpleComplex\Utils\Bootstrap;

/**
 * Slim HTTP service (route responder) base class.
 *
 * See the bootstrapper.
 * @see HttpServiceSlim::bootstrap()
 *
 * @uses-dependency-container locale, logger, inspect
 *
 * @package SimpleComplex\Http
 */
abstract class HttpServiceSlim
{
    /**
     * Final extending class _must_ override this.
     *
     * @var string
     */
    const APPLICATION_ID = 'application-id-unknown';

    /**
     * Dependency injection container ID.
     *
     * Final extending class _must_ override this.
     *
     * @var string
     */
    const DEPENDENCY_ID = 'http-service.unknown';

    /**
     * @var string
     */
    const CLASS_BOOTSTRAP = Bootstrap::class;

    /**
     * @var string
     */
    const CLASS_HTTP_SETTINGS = HttpSettings::class;

    /**
     * Final extending class _must_ override this.
     *
     * @var array
     */
    const ROUTES = [
        // Slim App route methods' second arg callable with single colon means:
        // - get before-colon-dependency and call it's after-colon-method
        [
            'http-method-lowercased', '/route', 'routeMethodName',
        ]
    ];

    /**
     * @var string
     */
    protected static $crossOriginSiteAllowed;

    /**
     * Declares Slim routes.
     *
     * Has to be static, otherwise redundant instantiation:
     * - request matches this class: double instantiation
     * - request matches other class: one (unneeded) instantiation
     *
     * Also declares OPTION routes, if any cross origin sites allowed.
     *
     * @param \Slim\App $app
     */
    public static function routes($app)
    {
        $cross_origin = !!(static::$crossOriginSiteAllowed = HttpService::crossOriginSiteAllowed());

        foreach (static::ROUTES as $route) {
            $methods = $route[0];
            if (is_array($methods)) {
                $app->map($methods, $route[1], static::DEPENDENCY_ID . ':' . $route[2]);
            }
            else {
                $app->{$methods}($route[1], static::DEPENDENCY_ID . ':' . $route[2]);
            }

            // Set OPTION routes to respond to cross origin 'pre-flight' OPTION
            // request, which a browser issues if a request sends custom headers.
            if ($cross_origin) {
                $app->options($route[1], static::DEPENDENCY_ID . ':crossOriginOptions');
            }
        }
    }

    /**
     * Provides application dependencies, now that we know which application
     * receives the request.
     */
    protected function __construct()
    {
        $container = Dependency::container();
        $application_id = static::APPLICATION_ID;
        $http_settings = static::CLASS_HTTP_SETTINGS;
        Dependency::genericSetMultiple(
            [
                // Make http class settings available.
                'http-settings' => function() use ($http_settings) {
                    return new $http_settings();
                },
                // Application specific.
                'application-id' => $application_id,
                'application-title' => function() use ($container, $application_id) {
                    // Use base application title as fallback,
                    // if the solution's locale text ini file misses
                    // [some-application-id]
                    // application-title = Some Solution.
                    // Non-solution services, will always use base title.
                    /** @var \SimpleComplex\Locale\AbstractLocale $locale */
                    $locale = $container->get('locale');
                    // Cascading: application-id or common or base.
                    return $locale->text(
                        $application_id . ':application-title',
                        [],
                        $locale->text(
                            'common:application-title',
                            [],
                            $locale->text('http:application-title')
                        )
                    );
                },
            ]
        );
    }

    /**
     * Send cache control response headers.
     *
     * See also Typescript angular.kk-base.base KkBaseAbstractHttpService.
     *
     * @param Response $response
     * @param int $timeToLive
     *      In seconds; default zero (prevent browser caching).
     *
     * @return Response
     *
     * @throws \InvalidArgumentException
     *      Propagated; arg timeToLive is negative.
     */
    public function cacheControl(Response $response, int $timeToLive = 0) : Response
    {
        $headers = HttpService::cacheControlHeaders($timeToLive);
        foreach ($headers as $key => $val) {
            $response = $response->withHeader($key, $val);
        }
        return $response;
    }

    /**
     * Make all but unauthorized 'forbidden' responses look the same.
     *
     * @see HttpService::RESPONSE_FORBIDDEN
     *
     * @param Response $response
     *
     * @return Response
     */
    public function respondForbidden(Response $response) : Response
    {
        /** @var HttpSettings $http_settings */
        $http_settings = Dependency::container()->get('http-settings');
        /** @var \Slim\Http\Response $response */
        $response = $response->withStatus($http_settings->serviceStatusCode('forbidden'));
        // Copy.
        $headers = $http_settings->serviceResponseForbidden();
        if (!empty($headers['body'])) {
            $response->write($headers['body']);
        }
        unset($headers['body']);
        foreach ($headers as $key => $val) {
            $response = $response->withHeader($key, $val);
        }
        return $response;
    }

    /**
     * Route method for responding to pre-flight OPTIONS request.
     *
     * @param Request $request
     * @param Response $response
     *
     * @return Response
     */
    public function crossOriginOptions(Request $request, Response $response)
    {
        // Allow http://localhost:4200 in development (ng serve, npm start),
        // using custom headers.
        return static::crossOriginOptionsSetHeaders($request, $response);
    }

    /**
     * Set Cross Origin headers, preferably only in development.
     *
     * Headers set:
     * - Access-Control-Allow-Origin; allow this sites
     * - Access-Control-Expose-Headers; allow that requestor reads
     *   those response headers
     * - Access-Control-Allow-Credentials; allow that requestor sends
     *   login credentials (like NTLM/Kerberos)
     *   Only if arg $requireCredentials.
     *
     * Circumvent CORS (cross origin resource sharing) check in development.
     * Angular serves from other host in development;
     * typically http://localhost:4200 (ng serve|npm start).
     *
     * Allowed sites are set as comma-separated list (including HTTP port)
     * in file [document root]/.cross_origin_allow_sites
     *
     * @see HttpService::crossOriginSiteAllowed()
     *
     * @param \Slim\Http\Response $response
     * @param bool $requireCredentials
     *      Must be true when using Kerberos authentication or likewise.
     *
     * @return \Slim\Http\Response
     */
    public static function crossOriginSetHeaders(Response $response, bool $requireCredentials = false) : Response
    {
        if (static::$crossOriginSiteAllowed) {
            // Allow requestor to see all relevant response headers;
            // headers already set plus Content-Length (which Slim sets
            // before returning response to requestor.
            $response = $response->withHeader(
                'Access-Control-Allow-Origin',
                static::$crossOriginSiteAllowed
            )->withHeader(
                'Access-Control-Expose-Headers',
                join(',', array_keys($response->getHeaders())) . ',Content-Length'
            );
            if ($requireCredentials) {
                $response = $response->withHeader(
                    'Access-Control-Allow-Credentials', 'true'
                );
            }
        }
        return $response;
    }

    /**
     * Append headers to Cross Origin exposable headers header.
     *
     * @param \Slim\Http\Response $response
     * @param array $responseHeaders
     *
     * @return Response
     */
    public static function crossOriginAppendHeaders(Response $response, array $responseHeaders) : Response {
        if ($responseHeaders && $response->hasHeader('Access-Control-Expose-Headers')) {
            // withHeader() overrides previously set headers by that name.
            $response = $response->withHeader(
                'Access-Control-Expose-Headers',
                join(
                    ',',
                    array_unique(
                        array_merge(
                            explode(',', $response->getHeader('Access-Control-Expose-Headers')[0]),
                            array_keys($responseHeaders)
                        )
                    )
                )
            );
        }
        return $response;
    }

    /**
     * For pre-flight OPTIONS request, sent by browser
     * typically due to custom request header.
     *
     * Headers set:
     * - Access-Control-Allow-Methods: all methods supported by our HTTP client,
     *   because we cannot know here which methods the endpoint supports
     * - Access-Control-Allow-Headers: same as the request's
     *   Access-Control-Request-Headers: list
     * - Access-Control-Max-Age: we don't want to see these requests too often.
     *
     * @param Request $request
     * @param Response $response
     *
     * @return Response
     */
    public static function crossOriginOptionsSetHeaders(
        Request $request,
        Response $response
    ) : Response {
        if (static::$crossOriginSiteAllowed) {
            // Set Access-Control-Allow-Origin, Access-Control-Expose-Headers.
            $response = static::crossOriginSetHeaders(
                $response,
                // Don't ever require authentication for OPTIONS request.
                false
            );
            // If the request contains custom headers, they will be listed
            // in request Access-Control-Request-Headers.
            if ($request->hasHeader('Access-Control-Request-Headers')) {
                // All HTTP methods supported by our HTTP client.
                /** @var Response $response */
                $response = $response->withHeader(
                    'Access-Control-Allow-Methods',
                    join(', ', HttpClient::methodsSupported())
                )->withHeader(
                    'Access-Control-Allow-Headers',
                    // Browser sends Access-Control-Request-Headers lowercased,
                    // and apparantly also understands response
                    // Access-Control-Allow-Headers lowercased.
                    join(', ', $request->getHeader('Access-Control-Request-Headers'))
                );
                $response = $response->withHeader(
                    'Access-Control-Max-Age',
                    '' . Dependency::container()->get('http-settings')->client('cacheable_time_to_live')
                );
            }
        }
        return $response;
    }

    /**
     * Bootstraps Slim, and the SimpleComplex framework for all services.
     *
     * @param Callable|null $customLogger
     *      Custom logger; default is JsonLog.
     *
     * @return \Slim\App
     */
    public static function bootstrap(/*?Callable*/ $customLogger = null)
    {
        // Create Slim dependency injection container.
        $container = new /*\Slim\*/Container;
        // Pass some settings to Slim.
        $container['settings']['displayErrorDetails'] = true;

        // Pass Slim container to SimpleComplex Dependency container.
        // The Slim container itself is still usable directly.
        // SimpleComplex classes use the container via Dependency,
        // to avoid dependency of a particular PSR Container (like Slim's).

        // And prepare base dependencies.
        /**
         * @see \SimpleComplex\Utils\Bootstrap::prepareDependencies()
         */
        $bootstrap = static::CLASS_BOOTSTRAP . '::prepareDependencies';
        $bootstrap($container, $customLogger);

        // Fallback exception handler.
        Bootstrap::setExceptionHandler($container, 'http');

        // PHP warning/notice handler; Slim doesn't handle those.
        Bootstrap::setErrorHandler($container, 'http');

        // Slim request PHP error handler.
        Dependency::genericSet('phpErrorHandler', function () use ($container) {
            return function (
                /*\Slim\Http*/Request $request,
                /*\Slim\Http*/Response $response,
                \Throwable $exception
            ) use ($container) {
                try {
                    // Log exception trace and request specifics.
                    $inspect = $container->get('inspect');
                    $container->get('logger')->error(
                        'PHP error'
                        . "\n" . $inspect->trace(null)
                        . "\n" . $inspect->variable([
                            'path' => $request->getUri()->getPath(),
                            'query' => $request->getQueryParams(),
                            'body' => $request->getParsedBody(),
                        ]),
                        [
                            'code' => $exception->getCode(),
                        ]
                    );
                } catch (\Throwable $xcptn) {
                    // Log original exception.
                    error_log(
                        get_class($exception) . '(' . $exception->getCode() . ')@' . $exception->getFile() . ':'
                        . $exception->getLine() . ': ' . addcslashes($exception->getMessage(), "\0..\37")
                    );
                    // Log this exception handler's own exception.
                    error_log(
                        get_class($xcptn) . '(' . $xcptn->getCode() . ')@' . $xcptn->getFile() . ':'
                        . $xcptn->getLine() . ': ' . addcslashes($xcptn->getMessage(), "\0..\37")
                    );
                }
                try {
                    // Even error response may need Cross Origin headers.
                    if (static::$crossOriginSiteAllowed) {
                        if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
                            $response = static::crossOriginOptionsSetHeaders($request, $response);
                        } else {
                            $response = static::crossOriginSetHeaders($response);
                        }
                    }
                } catch (\Throwable $ignore) {
                }
                /**
                 * Send status 200, pass real success+status in response body.
                 *
                 * @see \SimpleComplex\Http\HttpResponseBody
                 */
                $response = $response->withHeader('Content-Type', 'application/json');
                $response->write(
                    json_encode([
                        'success' => false,
                        'status' => 500,
                        // Empty: frontend will use locale text http-client:error:local-unknown.
                        'message' => '',
                        'data' => null,
                        'code' => 0,
                    ])
                );
                return $response
                    ->withStatus(200);
            };
        });
        // Slim request exception handler.
        Dependency::genericSet('errorHandler', function () use ($container) {
            return function (
                /*\Slim\Http*/Request $request,
                /*\Slim\Http*/Response $response,
                \Throwable $exception
            ) use ($container) {
                try {
                    // Log exception trace and request specifics.
                    $inspect = $container->get('inspect');
                    $container->get('logger')->error(
                        $inspect->trace($exception)
                        . "\n" . $inspect->variable([
                            'path' => $request->getUri()->getPath(),
                            'query' => $request->getQueryParams(),
                            'body' => $request->getParsedBody(),
                        ]),
                        [
                            'code' => $exception->getCode(),
                        ]
                    );
                } catch (\Throwable $xcptn) {
                    // Log original exception.
                    error_log(
                        get_class($exception) . '(' . $exception->getCode() . ')@' . $exception->getFile() . ':'
                        . $exception->getLine() . ': ' . addcslashes($exception->getMessage(), "\0..\37")
                    );
                    // Log this exception handler's own exception.
                    error_log(
                        get_class($xcptn) . '(' . $xcptn->getCode() . ')@' . $xcptn->getFile() . ':'
                        . $xcptn->getLine() . ': ' . addcslashes($xcptn->getMessage(), "\0..\37")
                    );
                }
                try {
                    // Even error response may need Cross Origin headers.
                    if (static::$crossOriginSiteAllowed) {
                        if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
                            $response = static::crossOriginOptionsSetHeaders($request, $response);
                        } else {
                            $response = static::crossOriginSetHeaders($response);
                        }
                    }
                } catch (\Throwable $ignore) {
                }
                /**
                 * Send status 200, pass real success+status in response body.
                 *
                 * @see \SimpleComplex\Http\HttpResponseBody
                 */
                $response = $response->withHeader('Content-Type', 'application/json');
                $response->write(
                    json_encode([
                        'success' => false,
                        'status' => 500,
                        // Empty: frontend will use locale text http:error-client:local-unknown.
                        'message' => '',
                        'data' => null,
                        'code' => $exception->getCode(),
                    ])
                );
                return $response
                    ->withStatus(200);
            };
        });

        // Init Slim application.
        return new /*\Slim\*/App($container);
    }
}
